### PR TITLE
trainee add edit display delete signin signout

### PR DESCRIPTION
**Pull request by MuhammadTamzid**
We've combined some related tasks in this pull request.

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
+      log_in user
+      params[:session][:remember_me] == '1' ? remember(user) : forget(user)
+      redirect_back_or user
+    else
+      flash.now[:danger] = 'Invalid email/password combination'
```
Comment (chuanhtuan) [inline app/controllers/sessions_controller.rb]: i18n

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
+class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update]
+  before_action :correct_user,   only: [:edit, :update]
+
+  def show_trainee
+    @users = User.where("role = ?", 3)
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: please write in a scope

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                                   :password_confirmation)
+    end
+
+    # Before filters
+
+    # Confirms a logged-in user.
+    def logged_in_user
+      unless logged_in?
+        store_location
+        flash[:danger] = "Please log in."
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: i18n

File: app/models/user.rb [hunk 4]
```diff
@@ -1,2 +1,45 @@
...
+
+  # Returns a random token.
+  def User.new_token
+    SecureRandom.urlsafe_base64
+  end
+
+  # Remembers a user in the database for use in persistent sessions.
+  def remember
+    self.remember_token = User.new_token
+    update_attribute(:remember_digest, User.digest(remember_token))
```
Comment (chuanhtuan) [inline app/models/user.rb]: use update_attributes instead of update_attribute

File: app/models/user.rb [hunk 5]
```diff
@@ -1,2 +1,45 @@
...
+  end
+
+  # Returns true if the given token matches the digest.
+  def authenticated? remember_token
+    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  end
+
+  # Forgets a user.
+  def forget
+    update_attribute(:remember_digest, nil)
```
Comment (chuanhtuan) [inline app/models/user.rb]: use update_attributes instead of update_attribute

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
+class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update]
+  before_action :correct_user,   only: [:edit, :update]
+
+  def show_trainee
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: remove
RESTful

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    @users = User.trainees.paginate page: params[:page]
+  end
+
+  def show_supervisor
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: remove
RESTful

File: config/routes.rb [hunk 8]
```diff
@@ -1,6 +1,15 @@
...
-  root 'static_pages#home'
-  get  'help'    => 'static_pages#help'
-  get  'about'   => 'static_pages#about'
-  get  'contact' => 'static_pages#contact'
+  scope "(:locale)", :locale => /en|vn/ do
+    root                           'static_pages#home'
+    get    'help'               => 'static_pages#help'
+    get    'about'              => 'static_pages#about'
+    get    'contact'            => 'static_pages#contact'
+    get    'add_trainee'        => 'users#new'
```
Comment (chuanhtuan) [inline config/routes.rb]: remove
RESTful

File: config/routes.rb [hunk 9]
```diff
@@ -1,6 +1,15 @@
...
+    get    'display_trainee'    => 'users#show_trainee'
```
Comment (chuanhtuan) [inline config/routes.rb]: remove
RESTful

File: config/routes.rb [hunk 10]
```diff
@@ -1,6 +1,15 @@
...
+    get    'display_supervisor' => 'users#show_supervisor'
```
Comment (chuanhtuan) [inline config/routes.rb]: remove
RESTful

File: config/routes.rb (hunk 8 above)
Comment (MuhammadTamzid) [inline config/routes.rb]: What to change here? 
Here we used "new" method

Comment (chuanhtuan) [inline config/routes.rb]: Ah, OK

File: app/views/layouts/_header.html.erb [hunk 11]
```diff
@@ -4,7 +4,48 @@
...
+          <li class="dropdown">
+            <a href="#" class="dropdown-toggle" data-toggle="dropdown">
+              Trainee <b class="caret"></b>
+            </a>
+            <ul class="dropdown-menu">
+              <% if is_supervisor? %>
+                <li><%= link_to "Add new", add_trainee_path %></li>
+              <% end %>
+              <li>
+                <%= link_to "Show all", users_path(:role => "trainee") %>
```
Comment (chuanhtuan) [inline app/views/layouts/_header.html.erb]: <%= link_to "Show all", users_path(role: "trainee") %>

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
+class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update]
+  before_action :correct_user,   only: [:edit, :update]
+
+  def index
+    @role = params[:role]
+    if @role == "trainee"
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: define a constant in model

File: app/models/user.rb [hunk 13]
```diff
@@ -1,2 +1,59 @@
...
+  before_save { self.email = email.downcase }
+  validates :name, presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 },
+                    format: { with: VALID_EMAIL_REGEX },
+                    uniqueness: { case_sensitive: false }
+  has_secure_password
+  validates :password, presence: true, length: { minimum: 6 }, allow_nil: true
+
+  enum role: [:superadmin, :admin, :supervisor, :trainee]
```
Comment (chuanhtuan) [inline app/models/user.rb]: only 2 role: :supervisor, :trainee

File: app/helpers/users_helper.rb [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+  def gravatar_for(user, options = { size: 160 })
+    gravatar_id = Digest::MD5::hexdigest user.email.downcase
+    size = options[:size]
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}?s=#{size}"
+    image_tag(gravatar_url, alt: user.name, class: "gravatar")
+  end
+
+  # Returns true if the user role is supervisor.
+  def is_supervisor?
+    current_user.role == "supervisor"
```
Comment (chuanhtuan) [inline app/helpers/users_helper.rb]: current_user.supervisor?

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -0,0 +1,68 @@
+class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update]
+  before_action :correct_user,   only: [:edit, :update]
+
+  def index
+    User.new.set_role params[:role]
+    if User::ROLE == "trainee"
+      @users = User.trainees.paginate page: params[:page]
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: User.new.send(params[:role].pluralize).paginate page: params[:page]
please try :D

File: app/models/user.rb [hunk 16]
```diff
@@ -1,2 +1,59 @@
...
+
+  scope :superadmins, -> {admin}
+  scope :admins, -> {admin}
+  scope :supervisors, -> {supervisor}
+  scope :trainees, -> {trainee}
+
+  ROLE = ""
```
Comment (chuanhtuan) [inline app/models/user.rb]: remove

File: app/views/users/new.html.erb [hunk 17]
```diff
@@ -0,0 +1,24 @@
+<h1><%= t(:add_trainee_text) %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
+      <%= render 'shared/error_messages', object: f.object %>
+
+      <%= f.hidden_field :role, :value => "trainee" %>
```
Comment (chuanhtuan) [inline app/views/users/new.html.erb]: remove
please set this attribute in controller

File: app/controllers/users_controller.rb (hunk 15 above)
Comment (MuhammadTamzid) [inline app/controllers/users_controller.rb]: Sir,
If we implement by this, there is an error occurred. "NoMethodError in UsersController#index" 
Do we have to implement "send" method in User model?

File: app/helpers/users_helper.rb (hunk 14 above)
Comment (MuhammadTamzid) [inline app/helpers/users_helper.rb]: If we this there is an error occurred, "undefined method `supervisor?' for "supervisor":String"

Comment (chuanhtuan) [inline app/helpers/users_helper.rb]: current_user.supervisor?

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,65 @@
+class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update]
+  before_action :correct_user,   only: [:edit, :update]
+
+  def index
+    @users = User.new.send(:get_users, params[:role].pluralize)
```
Comment (chuanhtuan) [inline app/controllers/users_controller.rb]: User.send(params[:role].pluralize)..paginate page: params[:page]

File: app/models/user.rb [hunk 19]
```diff
@@ -1,2 +1,60 @@
...
+  has_secure_password
+  validates :password, presence: true, length: { minimum: 6 }, allow_nil: true
+
+  enum role: [:supervisor, :trainee]
+
+  scope :supervisors, -> {supervisor}
+  scope :trainees, -> {trainee}
+
+  # Returns users by role
+  def get_users role
```
Comment (chuanhtuan) [inline app/models/user.rb]: remove

File: app/views/users/edit.html.erb [hunk 20]
```diff
@@ -0,0 +1,29 @@
+<% provide(:title, t(:update_profile_text)) %>
+<h1><%= t(:update_profile_text) %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
```
Comment (chuanhtuan) [inline app/views/users/edit.html.erb]: ()

File: app/views/users/new.html.erb [hunk 21]
```diff
@@ -0,0 +1,22 @@
+<h1><%= t(:add_trainee_text) %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
```
Comment (chuanhtuan) [inline app/views/users/new.html.erb]: ()

File: config/locales/vn.yml [hunk 22]
```diff
@@ -0,0 +1,10 @@
+vn:
+  add_trainee_sucessful_message: "Trainee thêm"
+  add_trainee_text: "Thêm Trainee"
+  login_text: "Đăng nhập"
```
Comment (chuanhtuan) [inline config/locales/vn.yml]: google, right? :D

Comment (MuhammadTamzid) [inline config/locales/vn.yml]: Yes, google translate :D

**Pull request merged**